File: River/River.Body/Ahoy/Messages/MessageReceiveSystem.py
from Ahoy.AhoyEvent import AhoyEvent
from Sockets.Systems import socketSystem
from Messages import MessageUtility, MessageEventArgs


class MessageReceiveSystem():
    def __init__(self):
        socketSystem.onData.AddListener(self.OnData)
        self.channelEvents = {}

    def GetOrCreateChannelReceiver(self, key):
        eve = self.channelEvents.get(key)
        if eve == None:
            eve = AhoyEvent()
        self.channelEvents[key] = eve
        return eve

    def AddChannelListener(self, channel, onMessage):
        eve = self.GetOrCreateChannelReceiver(channel)
        eve.AddListener(onMessage)

    def OnData(self, args):
        eve = self.channelEvents.get(args.data[0])
        if(eve == None):
            return
        # print("channel {}".format(args.data[0]))
        msg = MessageUtility.BytesToMessage(args.data)
        msgArgs = MessageEventArgs.MessageEventArgs(args.sender, args.receiver, msg)
        eve.Invoke(msgArgs)

File: River/River.Client/Assets/Packages/Ahoy.Net/Scripts/Messages/MessageReceiveSystem.cs

using System;
using System.Collections.Generic;
// using System.Linq;
// using System.Net;
// using System.Net.Socskets;
using System.Text;
using Unity.Entities;
// [Serializable]

using UnityEngine;
namespace Ahoy.Net
{

    using EventSet = Dictionary<byte, AhoyEvent<MessageEventArgs>>;
    // using ReceiveEvent =
    public class MessageReceiveSystem : ComponentSystem
    {

        public bool debug;

        public EventSet channelEvents;
        AhoySocketSystem socketSystem;

        protected override void OnCreate()
        {
            channelEvents = new EventSet();
            socketSystem = World.Active.GetOrCreateSystem<AhoySocketSystem>();
            socketSystem.onData.AddListener(OnData);
        }

        AhoyEvent<MessageEventArgs> GetOrCreateChannelReceiver(byte key)
        {
            AhoyEvent<MessageEventArgs> eve;
            if (!channelEvents.TryGetValue(key, out eve))
            {
                if (debug)
                    Debug.Log($"Message Receive System - creating event for channel {key}");
                // Debug.Log($"creating event");
                eve = new AhoyEvent<MessageEventArgs>();
                channelEvents[key] = eve;
            }
            return eve;
        }

        public void AddChannelListener(byte channel, Action<MessageEventArgs> onMessage)
        {
            var eve = GetOrCreateChannelReceiver(channel);
            eve.AddListener(onMessage);
        }

        public void OnData(SocketDataEventArgs args)
        {
            if (debug)
                Debug.Log($"Message Receive System - data received: {args}");
            if (args.data.Count < 1)
                return;
            AhoyEvent<MessageEventArgs> messageEvent;
            var channel = args.data.GetElement(0);
            if (channelEvents.TryGetValue(channel, out messageEvent))
            {
                var message = MessageParse.DecodeToMessage(args.data);
                var msgArgs = new MessageEventArgs(args.sender, args.receiver, message);
                if (debug)
                    Debug.Log($"Message Receive System - data event: {args}");
                messageEvent.Invoke(msgArgs);
            }
            else
            {
                if (debug)
                    Debug.Log($"Message Receive System - no channel receivers for channel {channel}: {args}");
            }
        }

        // public bool TrySetInstruction()
        // {


        // }

        protected override void OnUpdate() { }
        // public void TestCreate() { OnCreate(); }
    }


}